File: api/src/main/java/com/github/noonmaru/tap/collection/SortedList.java
/*
 * Copyright (c) 2020 Noonmaru
 *
 *  Licensed under the General Public License, Version 3.0 (the "License");
 *  you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://opensource.org/licenses/gpl-3.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.noonmaru.tap.collection;

import java.util.*;

public class SortedList<E> extends AbstractList<E> implements RandomAccess, Cloneable {

    final ArrayList<E> list;

    final Comparator<E> comparator;

    public SortedList() {
        this.list = new ArrayList<>();
        this.comparator = null;
    }

    public SortedList(Comparator<E> comparator) {
        this.list = new ArrayList<>();
        this.comparator = comparator;
    }

    public SortedList(int initialCapacity) {
        this.list = new ArrayList<>(initialCapacity);
        this.comparator = null;
    }

    public SortedList(int initialCapacity, Comparator<E> comparator) {
        this.list = new ArrayList<>(initialCapacity);
        this.comparator = comparator;
    }

    public SortedList(Collection<E> c) {
        this.list = new ArrayList<>(c);
        this.comparator = null;
        sort();
    }

    public SortedList(Collection<E> c, Comparator<E> comparator) {
        this.list = new ArrayList<>(c);
        this.comparator = comparator;
        sort();
    }

    @Override
    public boolean add(E element) {
        sort();
        binaryAdd(element);

        return true;
    }

    public void sort() {
        this.list.sort(this.comparator);
    }

    public int binaryAdd(E element) {
        if (element == null)
            throw new NullPointerException();

        this.modCount++;

        int size = list.size();

        if (size == 0) {
            list.add(element);
            return 0;
        }

        int index = binarySearch(element);

        if (index < 0)
            index = -(index + 1);

        list.add(index, element);

        return index;
    }

    public boolean binaryRemove(E o) {
        int index = binarySearch(o);

        if (index >= 0) {
            this.list.remove(index);

            return true;
        }

        return false;
    }

    @Override
    public boolean addAll(Collection<? extends E> c) {
        this.list.ensureCapacity(c.size());

        for (E e : c)
            add(e);

        return true;
    }

    @Override
    public void clear() {
        this.list.clear();
    }

    @Override
    public boolean contains(Object o) {
        return this.list.contains(o);
    }

    @Override
    public boolean containsAll(Collection<?> c) {
        return this.list.containsAll(c);
    }

    @SuppressWarnings("unchecked")
    @Override
    public boolean equals(Object o) {
        return o instanceof SortedList && this.list.equals(((SortedList<E>) o).list);
    }

    public int binarySearch(E element) {
        return Collections.binarySearch(this.list, element, this.comparator);
    }

    @Override
    public E get(int index) {
        return this.list.get(index);
    }

    @Override
    public int hashCode() {
        return this.list.hashCode();
    }

    @Override
    public int indexOf(Object o) {
        return this.list.indexOf(o);
    }

    @Override
    public boolean isEmpty() {
        return this.list.isEmpty();
    }

    @Override
    public int lastIndexOf(Object o) {
        return this.list.lastIndexOf(o);
    }

    @Override
    public ListIterator<E> listIterator(final int index) {
        return new ListIterator<E>() {
            private final ListIterator<E> iterator = SortedList.this.list.listIterator(index);

            @Override
            public void add(E e) {
                throw new UnsupportedOperationException();
            }

            @Override
            public boolean hasNext() {
                return this.iterator.hasNext();
            }

            @Override
            public boolean hasPrevious() {
                return this.iterator.hasPrevious();
            }

            @Override
            public E next() {
                return this.iterator.next();
            }

            @Override
            public int nextIndex() {
                return this.iterator.nextIndex();
            }

            @Override
            public E previous() {
                return this.iterator.previous();
            }

            @Override
            public int previousIndex() {
                return this.iterator.previousIndex();
            }

            @Override
            public void remove() {
                this.iterator.remove();
            }

            @Override
            public void set(E e) {
                throw new UnsupportedOperationException();
            }

        };
    }

    @Override
    public E remove(int index) {
        return this.list.remove(index);
    }

    @Override
    public boolean remove(Object o) {
        return this.list.remove(o);
    }

    @Override
    public boolean removeAll(Collection<?> c) {
        return this.list.removeAll(c);
    }

    @Override
    public boolean retainAll(Collection<?> c) {
        return this.list.retainAll(c);
    }

    @Override
    public int size() {
        return this.list.size();
    }

    @Override
    public List<E> subList(int fromIndex, int toIndex) {
        return Collections.unmodifiableList(super.subList(fromIndex, toIndex));
    }

    @Override
    public Object[] toArray() {
        return this.list.toArray();
    }

    @Override
    public <T> T[] toArray(T[] a) {
        return this.list.toArray(a);
    }

    @Override
    public String toString() {
        return this.list.toString();
    }

}

File: tap-api/src/main/kotlin/io/github/monun/tap/collection/SortedList.kt
/*
 * Tap
 * Copyright (C) 2021 Monun
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package io.github.monun.tap.collection

import java.util.*

@Suppress("unused")
class SortedList<E> : AbstractList<E>, RandomAccess, Cloneable {
    private val list: ArrayList<E>
    private val comparator: Comparator<E>?

    constructor() {
        list = ArrayList()
        comparator = null
    }

    constructor(comparator: Comparator<E>? = null) {
        list = ArrayList()
        this.comparator = comparator
    }

    constructor(initialCapacity: Int, comparator: Comparator<E>? = null) {
        list = ArrayList(initialCapacity)
        this.comparator = comparator
    }

    constructor(elements: Collection<E>, comparator: Comparator<E>? = null) {
        list = ArrayList(elements)
        this.comparator = comparator
        sort()
    }

    override fun add(element: E): Boolean {
        sort()
        binaryAdd(element)
        return true
    }

    fun sort() {
        Collections.sort(list, comparator)
    }

    fun binaryAdd(element: E): Int {
        modCount++
        val size = list.size
        if (size == 0) {
            list.add(element)
            return 0
        }
        var index = binarySearch(element)
        if (index < 0) index = -(index + 1)
        list.add(index, element)
        return index
    }

    fun binaryRemove(element: E): Boolean {
        val index = binarySearch(element)
        if (index >= 0) {
            list.removeAt(index)
            return true
        }
        return false
    }

    override fun addAll(elements: Collection<E>): Boolean {
        list.ensureCapacity(elements.size)
        for (element in elements) add(element)
        return true
    }

    override fun clear() {
        list.clear()
    }

    override operator fun contains(element: E): Boolean {
        return list.contains(element)
    }

    override fun containsAll(elements: Collection<E>): Boolean {
        return list.containsAll(elements)
    }

    override fun equals(other: Any?): Boolean {
        return other is SortedList<*> && list == other.list
    }

    private fun binarySearch(element: E): Int {
        return Collections.binarySearch(list, element, comparator)
    }

    override fun get(index: Int): E {
        return list[index]
    }

    override fun hashCode(): Int {
        return list.hashCode()
    }

    override fun indexOf(element: E): Int {
        return list.indexOf(element)
    }

    override fun isEmpty(): Boolean {
        return list.isEmpty()
    }

    override fun lastIndexOf(element: E): Int {
        return list.lastIndexOf(element)
    }

    override fun listIterator(index: Int): MutableListIterator<E> {
        return object : MutableListIterator<E> {
            private val iterator = list.listIterator(index)
            override fun add(element: E) {
                throw UnsupportedOperationException()
            }

            override fun hasNext(): Boolean {
                return iterator.hasNext()
            }

            override fun hasPrevious(): Boolean {
                return iterator.hasPrevious()
            }

            override fun next(): E {
                return iterator.next()
            }

            override fun nextIndex(): Int {
                return iterator.nextIndex()
            }

            override fun previous(): E {
                return iterator.previous()
            }

            override fun previousIndex(): Int {
                return iterator.previousIndex()
            }

            override fun remove() {
                iterator.remove()
            }

            override fun set(element: E) {
                throw UnsupportedOperationException()
            }
        }
    }

    override fun removeAt(index: Int): E {
        return list.removeAt(index)
    }

    override fun remove(element: E): Boolean {
        return list.remove(element)
    }

    override fun removeAll(elements: Collection<E>): Boolean {
        return list.removeAll(elements)
    }

    override fun retainAll(elements: Collection<E>): Boolean {
        return list.retainAll(elements)
    }

    override val size: Int
        get() {
            return list.size
        }

    override fun toArray(): Array<Any?> {
        return list.toTypedArray()
    }

    override fun <T> toArray(a: Array<T>): Array<T> {
        return list.toArray(a)
    }

    override fun toString(): String {
        return list.toString()
    }
}